File: lib/app/services/trading/matching_service_stop_loss_order_test.exs
defmodule MatchingServiceStopLossOrderTest do
  use DataCase

  import TestUtils

  test "create/1 stop loss sell order save" do
    # given
    account = acc()
    entity = TestUtils.get_application_entity_id()

    # when: a stop loss order is created
    res = MatchingService.create(account, "BTC_EUR", :STOPLOSS, :SELL, 10, 100, :GTC)

    # then:
    assert res != nil
    assert DBTestUtils.get_count("stop_order") == 1
    assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
    assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []

    assert PaymentAccount.find_by_application_entity_and_currency(entity, "BTC").amount_reserved
           |> Decimal.to_float() == 100
  end

  test "create/1 stop loss buy order buy" do
    # given
    account = acc()
    entity = TestUtils.get_application_entity_id()

    # when: a stop loss order is created
    res = MatchingService.create(account, "BTC_EUR", :STOPLOSS, :BUY, 10, 100, :GTC)

    # then:
    assert res != nil
    assert DBTestUtils.get_count("stop_order") == 1
    assert MatchingServiceTestHelpers.get_buy_book_order_count() == 0
    assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []

    assert PaymentAccount.find_by_application_entity_and_currency(entity, "EUR").amount_reserved
           |> Decimal.to_float() == 1000
  end

  test "create/1 stop loss sell order activate" do
    # given
    account = acc()
    account2 = acc2()

    # when: a stop loss order is created and then a crossing trade occurs
    MatchingService.create(account, "BTC_EUR", :STOPLOSS, :SELL, 10, 2, :GTC)
    MatchingService.create(account, "BTC_EUR", :LIMIT, :SELL, 10, 1, :GTC)
    MatchingService.create(account2, "BTC_EUR", :LIMIT, :BUY, 10, 1, :GTC)

    # then: the order becomes activated as a market order which is invvisible to the order book
    assert DBTestUtils.get_count("stop_order") == 0
    assert MatchingServiceTestHelpers.get_sell_book_order_count() == 1
    assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []
  end

  test "create/1 stop loss buy order activate" do
    # given
    account = acc()
    account2 = acc2()

    # when: a stop loss order is created and then a crossing trade occurs
    MatchingService.create(account, "BTC_EUR", :STOPLOSS, :BUY, 10, 2, :GTC)
    MatchingService.create(account2, "BTC_EUR", :LIMIT, :BUY, 10, 1, :GTC)
    MatchingService.create(account, "BTC_EUR", :LIMIT, :SELL, 10, 1, :GTC)

    # then: the order becomes activated as a market order which is invvisible to the order book
    assert DBTestUtils.get_count("stop_order") == 0
    assert MatchingServiceTestHelpers.get_buy_book_order_count() == 1
    assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
  end

  test "create/1 stop loss sell order non crossing " do
    # given
    account = acc()
    account2 = acc2()

    # when: a stop loss order is created and then a non-crossing trade occurs
    MatchingService.create(account, "BTC_EUR", :STOPLOSS, :SELL, 10, 2, :GTC)
    MatchingService.create(account, "BTC_EUR", :LIMIT, :SELL, 11, 1, :GTC)
    MatchingService.create(account2, "BTC_EUR", :LIMIT, :BUY, 11, 1, :GTC)

    # then: the order remains deactivated
    assert DBTestUtils.get_count("stop_order") == 1
    assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
    assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []

    # when: a crossing trade occurs
    MatchingService.create(account, "BTC_EUR", :LIMIT, :SELL, 9, 1, :GTC)
    MatchingService.create(account2, "BTC_EUR", :LIMIT, :BUY, 9, 1, :GTC)

    # then: it becomes activated
    assert DBTestUtils.get_count("stop_order") == 0
    assert MatchingServiceTestHelpers.get_sell_book_order_count() == 1
    assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []
  end

  test "create/1 stop loss buy order non crossing " do
    # given
    account = acc()
    account2 = acc2()

    # when: a stop loss order is created and then a non-crossing trade occurs
    MatchingService.create(account, "BTC_EUR", :STOPLOSS, :BUY, 10, 2, :GTC)
    MatchingService.create(account2, "BTC_EUR", :LIMIT, :BUY, 11, 1, :GTC)
    MatchingService.create(account, "BTC_EUR", :LIMIT, :SELL, 11, 1, :GTC)

    # then: the order remains deactivated
    assert DBTestUtils.get_count("stop_order") == 1
    assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
    assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []

    # when: a crossing trade occurs
    MatchingService.create(account2, "BTC_EUR", :LIMIT, :BUY, 9, 1, :GTC)
    MatchingService.create(account, "BTC_EUR", :LIMIT, :SELL, 9, 1, :GTC)

    # then: it becomes activated
    assert DBTestUtils.get_count("stop_order") == 0
    assert MatchingServiceTestHelpers.get_buy_book_order_count() == 1
    assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
  end

  test "create/1 stop loss sell order activate and settle" do
    # given
    account = acc()
    account2 = acc2()

    # when: a stop loss order is created and then a crossing trade occurs
    MatchingService.create(account, "BTC_EUR", :STOPLOSS, :SELL, 10, 2, :GTC)
    MatchingService.create(account, "BTC_EUR", :LIMIT, :SELL, 10, 1, :GTC)
    MatchingService.create(account2, "BTC_EUR", :LIMIT, :BUY, 10, 3, :GTC)

    # then: the order becomes activated and settled
    assert DBTestUtils.get_count("stop_order") == 0
    assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
    assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
    assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []
  end

  test "create/1 stop loss buy order activate and settle" do
    # given
    account = acc()
    account2 = acc2()

    # when: a stop loss order is created and then a crossing trade occurs
    MatchingService.create(account, "BTC_EUR", :STOPLOSS, :BUY, 10, 20, :GTC)
    MatchingService.create(account, "BTC_EUR", :LIMIT, :BUY, 10, 1, :GTC)
    MatchingService.create(account2, "BTC_EUR", :LIMIT, :SELL, 10, 3, :GTC)

    # then: the order becomes activated and settled
    assert DBTestUtils.get_count("stop_order") == 0
    assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
    assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
    assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []
  end

  # test "create/1 stop loss sell order activate and settle opposite side" do
  #   # given
  #   account = acc()
  #   account2 = acc2()

  #   # when: a stop loss order is created and then a crossing trade occurs
  #   MatchingService.create(account, "BTC_EUR", :STOPLOSS, :SELL, 10, 2, :GTC)
  #   MatchingService.create(account2, "BTC_EUR", :LIMIT, :BUY, 10, 3, :GTC)
  #   MatchingService.create(account, "BTC_EUR", :LIMIT, :SELL, 10, 1, :GTC)

  #   # then: the order becomes activated and settled
  #   assert DBTestUtils.get_count("stop_order") == 0
  #   assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
  #   assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
  #   assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []
  # end

  # test "create/1 stop loss buy order activate and settle opposite side" do
  #   # given
  #   account = acc()
  #   account2 = acc2()

  #   # when: a stop loss order is created and then a crossing trade occurs
  #   MatchingService.create(account, "BTC_EUR", :STOPLOSS, :BUY, 10, 20, :GTC)
  #   MatchingService.create(account2, "BTC_EUR", :LIMIT, :SELL, 10, 3, :GTC)
  #   MatchingService.create(account, "BTC_EUR", :LIMIT, :BUY, 10, 1, :GTC)

  #   # then: the order becomes activated and settled
  #   assert DBTestUtils.get_count("stop_order") == 0
  #   assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
  #   assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
  #   assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []
  # end

  test "create/1 stop loss sell order activate and settle before worse price orders" do
    # given
    account = acc()
    account2 = acc2()

    # when: a stop loss order is created and then a crossing trade occurs
    MatchingService.create(account, "BTC_EUR", :STOPLOSS, :SELL, 10, 1, :GTC)
    MatchingService.create(account, "BTC_EUR", :LIMIT, :SELL, 11, 1, :GTC)
    MatchingService.create(account, "BTC_EUR", :LIMIT, :SELL, 10, 1, :GTC)
    MatchingService.create(account2, "BTC_EUR", :LIMIT, :BUY, 11, 3, :GTC)

    # then: the order becomes activated and settled at last trade price which is 10
    assert MatchingServiceTestHelpers.get_trade_prices() == [10, 10, 11]
    assert DBTestUtils.get_count("stop_order") == 0
    assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
    assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
    assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []
  end

  test "create/1 stop loss buy order activate and settle before worse price orders" do
    # given
    account = acc()
    account2 = acc2()

    # when: a stop loss order is created and then a crossing trade occurs
    MatchingService.create(account, "BTC_EUR", :STOPLOSS, :BUY, 10, 10, :GTC)
    MatchingService.create(account, "BTC_EUR", :LIMIT, :BUY, 9, 1, :GTC)
    MatchingService.create(account, "BTC_EUR", :LIMIT, :BUY, 10, 1, :GTC)
    MatchingService.create(account2, "BTC_EUR", :LIMIT, :SELL, 9, 3, :GTC)

    # then: the order becomes activated and settled
    assert MatchingServiceTestHelpers.get_trade_prices() == [10, 10, 9]
    # TODO fix rounding
    assert DBTestUtils.get_count("stop_order") == 0
    assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
    assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
    assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []
  end

  # test "create/1 stop loss sell order activate by market and settle" do
  #   # given
  #   account = acc()
  #   account2 = acc2()

  #   # when: a stop loss order is created and then a crossing trade occurs
  #   MatchingService.create(account, "BTC_EUR", :STOPLOSS, :SELL, 10, 2, :GTC)
  #   MatchingService.create(account, "BTC_EUR", :MARKET, :SELL, 1, :GTC)
  #   MatchingService.create(account2, "BTC_EUR", :LIMIT, :BUY, 10, 3, :GTC)

  #   # then: the order becomes activated and settled
  #   assert DBTestUtils.get_count("stop_order") == 0
  #   assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
  #   assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
  #   assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []

  #   # when: a stop loss order is created and then a crossing trade occurs
  #   MatchingService.create(account, "BTC_EUR", :STOPLOSS, :SELL, 10, 2, :GTC)
  #   MatchingService.create(account2, "BTC_EUR", :LIMIT, :BUY, 10, 3, :GTC)
  #   MatchingService.create(account, "BTC_EUR", :MARKET, :SELL, 1, :GTC)

  #   # then: the order becomes activated and settled
  #   assert DBTestUtils.get_count("stop_order") == 0
  #   assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
  #   assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
  #   assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []
  # end

  # test "create/1 stop loss buy order activate by market and settle" do
  #   # given
  #   account = acc()
  #   account2 = acc2()

  #   # when: a stop loss order is created and then a crossing trade occurs
  #   MatchingService.create(account, "BTC_EUR", :STOPLOSS, :BUY, 10, 20, :GTC)
  #   MatchingService.create(account, "BTC_EUR", :MARKET, :BUY, 10, :GTC)
  #   MatchingService.create(account2, "BTC_EUR", :LIMIT, :SELL, 10, 3, :GTC)

  #   # then: the order becomes activated and settled
  #   assert DBTestUtils.get_count("stop_order") == 0
  #   assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
  #   assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
  #   assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []

  #   # when: a stop loss order is created and then a crossing trade occurs
  #   MatchingService.create(account, "BTC_EUR", :STOPLOSS, :BUY, 10, 20, :GTC)
  #   MatchingService.create(account2, "BTC_EUR", :LIMIT, :SELL, 10, 3, :GTC)
  #   MatchingService.create(account, "BTC_EUR", :MARKET, :BUY, 10, :GTC)

  #   # then: the order becomes activated and settled
  #   assert DBTestUtils.get_count("stop_order") == 0

  #   assert MatchingServiceTestHelpers.get_sell_book_order_count() == 0
  #   assert OrderBookService.get_volumes("BTC_EUR", :BUY) == []
  #   assert OrderBookService.get_volumes("BTC_EUR", :SELL) == []
  # end
end

File: pkg/services/matching_service_stop_loss_order_test.go
package services

import (
	"open-outcry/pkg/models"
	"open-outcry/pkg/utils"
)

func (assert *ServiceTestSuite) TestCreateStopLossSellOrderSave() {

	// when: a stop loss order is created
	res, _ := ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Sell, 10, 100, "GTC")

	// then:
	assert.NotNil(res)
	assert.Equal(1, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))

	assert.Equal(100.00, models.FindPaymentAccountByAppEntityIdAndCurrencyName(assert.appEntity1, "BTC").AmountReserved)
}

func (assert *ServiceTestSuite) TestCreateStoLossBuyOrderBuy() {

	// when: a stop loss order is created
	res, _ := ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Buy, 10, 100, "GTC")

	// then:
	assert.NotNil(res)
	assert.Equal(1, utils.GetCount("stop_order"))
	assert.Equal(0, GetBuyBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
	assert.Equal(1000.00, models.FindPaymentAccountByAppEntityIdAndCurrencyName(assert.appEntity1, "EUR").AmountReserved)
}
func (assert *ServiceTestSuite) TestCreateStopLossSellOrderActivate() {

	// when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Sell, 10, 2, "GTC")

	// then
	assert.Equal(2.0, models.FindPaymentAccountByAppEntityIdAndCurrencyName(assert.appEntity1, "BTC").AmountReserved)

	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Sell, 10, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Buy, 10, 1, "GTC")

	// then: the order becomes activated as a market order which is invisible to the order book
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(1, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))

	// activation has no affect on AmountReserved
	assert.Equal(2.0, models.FindPaymentAccountByAppEntityIdAndCurrencyName(assert.appEntity1, "BTC").AmountReserved)
}

func (assert *ServiceTestSuite) TestCreateStopLosBuyOrderActivate() {

	// when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Buy, 10, 2, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Buy, 10, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Sell, 10, 1, "GTC")

	assert.Equal(20.0, models.FindPaymentAccountByAppEntityIdAndCurrencyName(assert.appEntity1, "EUR").AmountReserved)

	// then: the order becomes activated as a market order which is invisible to the order book
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(1, GetBuyBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))

	assert.Equal(20.0, models.FindPaymentAccountByAppEntityIdAndCurrencyName(assert.appEntity1, "EUR").AmountReserved)
}

func (assert *ServiceTestSuite) TestCreateStopLossSellOrderNonCrossing() {

	// when: a stop loss order is created and then a non-crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Sell, 10, 2, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Sell, 11, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Buy, 11, 1, "GTC")

	// then: the order remains deactivated
	assert.Equal(1, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))

	// when: a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Sell, 9, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Buy, 9, 1, "GTC")

	// then: it becomes activated
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(1, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))
}

func (assert *ServiceTestSuite) TestCreateStopLossBuyOrderNonCrossing() {

	// when: a stop loss order is created and then a non-crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Buy, 10, 2, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Buy, 11, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Sell, 11, 1, "GTC")

	// then: the order remains deactivated
	assert.Equal(1, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))

	// when: a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Buy, 9, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Sell, 9, 1, "GTC")

	// then: it becomes activated
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(1, GetBuyBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
}

func (assert *ServiceTestSuite) TestCreateStopLossSellOrderActivateAndSettle() {

	// when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Sell, 10, 2, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Sell, 10, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Buy, 10, 3, "GTC")

	// then: the order becomes activated and settled
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))
}

func (assert *ServiceTestSuite) TestCreateStopLossBuyOrderActivateAndSettle() {

	// when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Buy, 10, 20, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Buy, 10, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Sell, 10, 3, "GTC")

	// then: the order becomes activated and settled
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))
}

func (assert *ServiceTestSuite) TestCreateStopLossSellOrderActivateAndSettleOppositeSide() {

	// when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Sell, 10, 2, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Buy, 10, 3, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Sell, 10, 1, "GTC")

	// then: the order becomes activated and settled
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))
}

func (assert *ServiceTestSuite) TestCreateStopLossBuyOrderActivateAndSettleOppositeSide() {

	// when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Buy, 10, 20, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Sell, 10, 3, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Buy, 10, 1, "GTC")

	// then: the order becomes activated and settled
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))
}

func (assert *ServiceTestSuite) TestCreateStopLossSellOrderActivateAndSettleBeforeWorsePriceOrders() {

	// when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Sell, 10, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Sell, 11, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Sell, 10, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Buy, 11, 3, "GTC")

	// then: the order becomes activated and settled at last trade price which is 10
	assert.Equal([]float64{10.0, 10.0, 11.0}, GetTradePrices())
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))
}

func (assert *ServiceTestSuite) TestCreateStopLossBuyOrderActivateAndSettleBeforeWorsePriceOrders() {

	// when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Buy, 10, 10, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Buy, 9, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", "LIMIT", models.Buy, 10, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Sell, 9, 3, "GTC")

	// then: the order becomes activated and settled
	assert.Equal([]float64{10.0, 10.0, 9.0}, GetTradePrices())
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))
}

func (assert *ServiceTestSuite) TestCreateStopLossSellOrderActivateByMarketAndSettle() {

	// when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Sell, 10, 2, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.Market, models.Sell, 0, 1, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Buy, 10, 3, "GTC")

	// then: the order becomes activated and settled
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))

	// when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Sell, 10, 2, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Buy, 10, 3, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.Market, models.Sell, 0, 1, "GTC")

	// then: the order becomes activated and settled
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))
}

func (assert *ServiceTestSuite) TestCreateStopLossBuyOrderActivateByMarketAndSettle() {

	// when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Buy, 10, 20, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.Market, models.Buy, 0, 10, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Sell, 10, 3, "GTC")

	// then: the order becomes activated and settled
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))

	//  when: a stop loss order is created and then a crossing trade occurs
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.StopLoss, models.Buy, 10, 20, "GTC")
	ProcessTradeOrder(assert.tradingAccount2, "BTC_EUR", "LIMIT", models.Sell, 10, 3, "GTC")
	ProcessTradeOrder(assert.tradingAccount1, "BTC_EUR", models.Market, models.Buy, 0, 10, "GTC")

	// then: the order becomes activated and settled
	assert.Equal(0, utils.GetCount("stop_order"))
	assert.Equal(0, GetSellBookOrderCount())
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Buy))
	assert.Equal([]PriceVolume{}, GetVolumes("BTC_EUR", models.Sell))
}
